### ISSUE TITLE
Problems with `survival::coxph()` and env handling

### ISSUE THREAD

**Issue by rkb965** [tidymodels/butcher#260]
## perhaps reporting if any changes are made, even if overall object size is unchanged?

(thank you for a wonderful package and apologies if any of the below is due to my own misunderstandings!)

It seems to me that verbose = TRUE reports an unsuccessful `butcher` if the total memory of an object does not change, even though aspects of the object may successfully be butchered (I will admit to having only a very fuzzy understanding of memory).

I was confused by output when I ran into this situation and initially thought that `butcher` was not able to trim my coxph object:

``` r
library(tibble)
library(lobstr)
library(survival)
library(butcher)

example_data <-
  tibble::tibble(
    time = rpois(10000, 2) + 1,
    status = rbinom(10000, 1, .5),
    x = rpois(10000, .5),
    covar = rbinom(10000, 1, .5)
  )

coxph_formula <- function(data) {
  res <- coxph(Surv(time, status) ~ x + strata(covar), data)
  res_formula <- res$formula
}

coxph_formula_butchered <- function(data) {
  res <- coxph(Surv(time, status) ~ x + strata(covar), data)
  cleaned_res <- butcher(res, verbose = TRUE)
  res_formula <- cleaned_res$formula
}

res <- coxph_formula(example_data)
res_butchered <- coxph_formula_butchered(example_data)
#> ✖ No memory released. Do not butcher.

lobstr::obj_size(res)
#> 534.84 kB
lobstr::obj_size(res_butchered)
#> 1.49 kB
```

<sup>Created on 2023-04-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

but clearly the formula object is much reduced. I am assuming that the env (or the contents of the env) that is removed from `formula` exists in a different object, so the total size of `res_butchered` is unchanged and thus `assess_object` determines that nothing net changed. But would it be possible to report `memory_released` on the level of the implemented `axe` s as well as the overall object size?

(The message changes to the expected information about memory being released if only I had butchered the thing I was truly interested in (`res_formula <- butcher(res$formula, verbose = TRUE)`) rather than turning it into a two-step process as I did, but memory is confusing and a more informative message could have prevented a few of the rabbit holes I went down before this all made sense)

Thanks for considering this!

**Comment by juliasilge** [comment]
Oh, something weird is going on with this:

``` r
library(tibble)
library(survival)
library(butcher)

example_data <-
  tibble::tibble(
    time = rpois(10000, 2) + 1,
    status = rbinom(10000, 1, .5),
    x = rpois(10000, .5),
    covar = rbinom(10000, 1, .5)
  )

## no function:
res <- coxph(Surv(time, status) ~ x + strata(covar), example_data)
butchered <- butcher(res, verbose = TRUE)
#> ✔ Memory released: "161.24 kB"

lobstr::obj_size(res)
#> 333.50 kB
lobstr::obj_size(butchered)
#> 172.26 kB


## inside of function
coxph_formula_butchered <- function(data) {
  res <- coxph(Surv(time, status) ~ x + strata(covar), data)
  butchered <- butcher(res, verbose = TRUE)
  butchered
}

res_via_func <- coxph_formula_butchered(example_data)
#> ✖ No memory released. Do not butcher.
```

<sup>Created on 2023-04-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by juliasilge** [comment]
This is a weird one, for sure. I think it is partly because of the specific use of the `terms` slot, which @simonpcouch pointed out in #243 is used in the predict method. We can use the debugger to drop into some of the functions to see the weirdness.

This is what you see with `debugonce(add_butcher_attributes)`:

```r
Browse[2]> lobstr::obj_size(old)
1.17 MB
Browse[2]> lobstr::obj_size(x)
1.18 MB
```
Butchered version is _bigger_? 👀 

I see the same thing with `debugonce(memory_released)` and `lobstr::obj_size()`, but then this result doesn't make much sense at all:

```r
Browse[2]> weigh(og)
# A tibble: 20 × 2
   object                    size
   <chr>                    <dbl>
 1 terms                 1.18    
 2 formula               1.18    
 3 residuals             0.722   
 4 y                     0.162   
 5 linear.predictors     0.0800  
 6 call                  0.0014  
 7 concordance           0.000752
 8 coefficients          0.00028 
 9 means                 0.00028 
10 wald.test             0.00028 
11 var                   0.000224
12 xlevels.strata(covar) 0.000176
13 method                0.000112
14 loglik                0.000064
15 score                 0.000056
16 iter                  0.000056
17 n                     0.000056
18 nevent                0.000056
19 assign.x              0.000056
20 timefix               0.000056
Browse[2]> weigh(butchered)
# A tibble: 20 × 2
   object                    size
   <chr>                    <dbl>
 1 terms                 1.18    
 2 residuals             0.722   
 3 linear.predictors     0.0800  
 4 formula               0.00149 
 5 call                  0.0014  
 6 concordance           0.000752
 7 coefficients          0.00028 
 8 means                 0.00028 
 9 wald.test             0.00028 
10 var                   0.000224
11 xlevels.strata(covar) 0.000176
12 method                0.000112
13 loglik                0.000064
14 score                 0.000056
15 iter                  0.000056
16 n                     0.000056
17 nevent                0.000056
18 assign.x              0.000056
19 timefix               0.000056
20 y                     0.000048
```

Maybe the `terms` and `formula` slots use the same environment? Yes:

``` r
library(tibble)
library(survival)
library(butcher)

example_data <-
  tibble::tibble(
    time = rpois(10000, 2) + 1,
    status = rbinom(10000, 1, .5),
    x = rpois(10000, .5),
    covar = rbinom(10000, 1, .5)
  )

res <- coxph(Surv(time, status) ~ x + strata(covar), example_data)
butchered <- butcher(res, verbose = TRUE)
#> ✔ Memory released: "161.33 kB"

lobstr::obj_addr(attr(res$formula, ".Environment"))
#> [1] "0x13e8452c8"
lobstr::obj_addr(attr(res$terms, ".Environment"))
#> [1] "0x13e8452c8"
```

<sup>Created on 2023-04-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

So I guess the current butchering methods clean out the environment from one slot but leave it for the other slot, depending on where it is relative to the global environment? Meaning nothing actually is getting removed at all from the object in total? Is that right -- `axe_env` sometimes does nothing?

It seems like sometimes `axe_data` sometimes does nothing too, in my exploration, probably because the `"y"` object we try to switch out _isn't a numeric vector at all_ but probably something also carrying around an env somehow:

``` r
library(tibble)
library(survival)

example_data <-
  tibble::tibble(
    time = rpois(10000, 2) + 1,
    status = rbinom(10000, 1, .5),
    x = rpois(10000, .5),
    covar = rbinom(10000, 1, .5)
  )

res <- coxph(Surv(time, status) ~ x + strata(covar), example_data)
class(res$y)
#> [1] "Surv"
```

<sup>Created on 2023-04-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

So (congrats to anyone who read through all this) what are our options, given that it looks to me like there is no way to axe components except when the model is in the global environment?

- Leave the butcher methods as is, because most people work in the global environment.
- Remove the butcher methods, because they will behave in surprising ways when used inside of functions (including possibly increasing model size).

**Comment by rkb965** [comment]
Hi Julia,

Thanks so much for diving into this! I assume your questions were mostly intended for your wise collaborators on this package and not for people who just learned of its existence three days ago, but:

> So I guess the current butchering methods clean out the environment from one slot but leave it for the other slot, depending on where it is relative to the global environment? Meaning nothing actually is getting removed at all from the object in total? Is that right -- axe_env sometimes does nothing?

&

> what are our options, given that it looks to me like there is no way to axe components except when the model is in the global environment?

I would disagree with the characterization that "nothing getting removed from the obj in total" is the same as "doing nothing." In my case, cleaning out the environment from one slot (conveniently, the slot I was later holding on to) *massively* reduced the size of my objects, and I am so glad that you provide the functionality to do that! 

Also, 

> Leave the butcher methods as is, because most people work in the global environment.

(I don't think wrapping a Cox model within a function is a particularly odd use case, and that's all it takes to get out of the global environment.)

Suggestion:

I actually still think that a slightly different behavior when verbose = TRUE would provide sufficient information to users to make sense of these somewhat confusing (by virtue of model objects nuances and environments) functions.

I very conveniently learned about `butcher` from someone who was both quite familiar with the package and also very patient and amenable to many follow-up questions about butchering and environments, so I kept digging until these nuances about shared environments, etc, made sufficient sense to me. But in another (more likely) universe where I discovered `butcher` on my own, I expect my experience would have been:

- Stage 1 (years of analysis, no understanding of objects, size, or environments): "Cox objects are sometimes kind of large"
- Stage 2 (`butcher` discovery, with "No memory released. Do not butcher." as above): "Huh. Cox objects can be large and I guess there is nothing to be done about it"

whereas with a more informative verbose = TRUE, I would have instead entered stage 2b: "Huh. Cox objects do not necessarily need to be this large and I need to dig into these `butcher` and `lobstr` details to better understand what I truly want to keep."

Suggestion:
I think that verbose = TRUE could print a few pieces of information that would be helpful for these atypical objects: 
(1) which `axe` methods were attempted (this https://butcher.tidymodels.org/articles/available-axe-methods.html is nice but in a bit of documentation depths)
(2) what success was had at the level of each `axe` method, with a succinct mention that memory and objects can be nuanced, and memory released at levels of `axe` may not be be released overall
(3) what success was had at the level of the total object (what is currently reported)

I changed these lines:
https://github.com/tidymodels/butcher/blob/71b1ae9993d88fb24f89884d35c4d703fa7f2f95/R/axe-generics.R#L18-L24

to verbose = TRUE and re-ran my earlier reprex. The output was a bit confusing because it's not immediately clear which lines correspond with which method, but after a little formatting, I think would actually be very informative: It tells me "some reductions can be made to your object size on some level (and maybe you should dive in and learn about these details, because it may be on a level that you care about)"

> res_butchered <- coxph_formula_butchered(example_data)
✔ Memory released: "0 B"
✖ Could not add <butchered> class
✔ Memory released: "0 B"
✖ Could not add <butchered> class
✖ No memory released. Do not butcher.
✔ Memory released: "533.24 kB"
✖ No memory released. Do not butcher.
✔ Memory released: "0 B"
✖ No memory released. Do not butcher.

Again, thanks for the package and thanks for your consideration! I wish I had the ability to just PR this suggestion, but that is not yet within my skillset.

**Comment by simonpcouch** [comment]
A super informative + thoughtful thread!🍄

I agree that the output:

```
res_via_func <- coxph_formula_butchered(example_data)
#> ✖ No memory released. Do not butcher.
```

is misleading.

<details>
  <summary>It's so wild that there's something specific about the global env, and not just a cluttered env generally? What?🫣</summary>
  
``` r
library(tibble)
library(survival)
library(butcher)

example_data <-
  tibble::tibble(
    time = rpois(10000, 2) + 1,
    status = rbinom(10000, 1, .5),
    x = rpois(10000, .5),
    covar = rbinom(10000, 1, .5)
  )

coxph_formula_butchered <- function(data) {
  # make a big ol object:
  boop <- rnorm(1e6)
  
  # do the rest as before:
  res <- coxph(Surv(time, status) ~ x + strata(covar), data)
  butcher(res, verbose = TRUE)
}

res_via_func <- coxph_formula_butchered(example_data)
#> ✖ No memory released. Do not butcher.
```

<sup>Created on 2023-04-26 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

</details>

Some stray thoughts:

- coxph objects are _super tricky_. The usage of environments and terms at predict time is quite unusual, and thus our ability to butcher them is somewhat hampered, too. In some ways, I feel like the output of `butcher(..., verbose = TRUE)` in this case just hits some edge cases that aren't relevant for most axe methods, and changing the generic for `butcher()` to accommodate that case (setting `verbose = TRUE` in each of the `axe()` methods inside of `butcher()`, especially) might be to the detriment of other use cases.

- That said, `memory_released()` might be able to be a bit more nuanced. We could raise a different message than "[n]o memory released" if `sum(weigh(og)$size) < sum(weigh(butcher)$size))`, though I'm not sure how commonly that happens / how many false positives we'd run in to. In that case, the message could read something like "butcher reduced the size of some slots of the object, but no memory was released in total." with a link to a help-file that gives R memory a slightly more thorough treatment.

- I'm not sure whether your stage 2 or 2b, @rkb965, is a more accurate portrayal of what's going on here, largely because I haven't interacted with coxph objects much beyond working on those butcher methods. It sounds like the reduction in memory seen in that formula env was actually helpful for you? Were you at some point able to use only the standalone formula object (or the model object without its `terms`) and actually free up some memory? (It may be worth looping in Hannah here, too, @juliasilge!)

- I believe the increase in memory is due to the class added on to `formula`.😱 Debugging inside of `axe_env.coxph()`:

```
Browse[3]> attributes(res$formula)
$class
[1] "butchered_formula" "formula"          

$.Environment
<environment: base>

Browse[3]> attributes(x$formula)
$class
[1] "formula"

$.Environment
<environment: R_GlobalEnv>

Browse[3]> lobstr::obj_size(attr(res$formula, "class"))
200 B
Browse[3]> lobstr::obj_size(attr(x$formula, "class"))
112 B
Browse[3]> lobstr::obj_size(attr(res$formula, ".Environment"))
0 B
Browse[3]> lobstr::obj_size(attr(x$formula, ".Environment"))
0 B
```

For that to actually account for the 1.17 MB -> 1.18 MB would need quite the coincidence for rounding, but...

> So I guess the current butchering methods clean out the environment from one slot but leave it for the other slot...

Yes!

> depending on where it is relative to the global environment? 

The butcher methods don't know anything about the global env specifically, but coxph seems to.

> Meaning nothing actually is getting removed at all from the object in total?  

Yes! In the situation where the model is fitted somewhere other than the global env, taken as a whole, the object is not made smaller due to the substitution of the formula env. The formula itself, though, is made smaller.

I think my recommendation would be to keep the coxph axe methods, leave the `butcher()` generic alone, and explore refining `memory_released()`'s conditions for raising the "[n]o memory released" message.

**Comment by juliasilge** [comment]
> It's so wild that there's something specific about the global env, and not just a cluttered env generally? What?🫣

YES 😬

I chatted with @hfrick and she shared that thankfully this works when wrapping in parsnip/censored:

``` r
library(tibble)
library(survival)
library(butcher)
library(censored)
#> Loading required package: parsnip

example_data <-
  tibble::tibble(
    time = rpois(10000, 2) + 1,
    status = rbinom(10000, 1, .5),
    x = rpois(10000, .5),
    covar = rbinom(10000, 1, .5)
  )

cox_censored <- proportional_hazards() %>%
  fit(Surv(time, status) ~ x + strata(covar), example_data)
censored_butchered <- butcher(cox_censored, verbose = TRUE)
#> ✔ Memory released: "160.66 kB"

lobstr::obj_size(cox_censored)
#> 1.17 MB
lobstr::obj_size(censored_butchered)
#> 1.01 MB
```

<sup>Created on 2023-04-26 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Maybe the use of quosures makes it work OK? 🫠

Thanks for your ideas @rkb965; I think our next steps will be:

- keep the coxph axe methods
- leave the `butcher()` generic alone
- change the message for `sum(weigh(og)$size) < sum(weigh(butcher)$size))` in `memory_released()`
- add disclaimer/advice to the docs for coxph ([this page](https://butcher.tidymodels.org/reference/axe-coxph.html)) to not butcher if you have wrapped in a function

**Comment by rkb965** [comment]
I 100% understand that this is completely resolved and closed (wow, y'all are fast!! impressive! thank you!) and am just responding to @simonpcouch 's question for completeness

>  It sounds like the reduction in memory seen in that formula env was actually helpful for you? Were you at some point able to use only the standalone formula object (or the model object without its terms) and actually free up some memory? 

Yes!! 

I was not trying to predict, just trying to save somewhat thorough output for completeness/inspection.

Without trying to bore you with too many details, I am working with changing data and had written functions to deal with various details, each of which contained a few coxph objects. I initially saved the entire coxph objects, having absolutely no idea how many copies of various environments I had in there. But after mapping over some things, I saved an 80 GB object and realized something might be amiss :)

Enter butcher:
I really only wanted to keep track of two things: estimated parameters and the formula that created them. The adjustment variables for various models could vary some and there was enough abstraction that I thought the most straightforward way to grab the details of the model formula was to grab the the coxph formula object (HA) which I thought was a character string, or something like it. Ah, sooo naive. But, saving just a broom::tidy'd coxph object result as well as a butchered coxph formula led me back down to very happy size-land (~1 MB, iirc). I tidy'd and butchered within my functions, and I quite liked the results.

After seeing this discussion today, I did personally re-write my use case to just keep track of the model formula (the character string that can become a formula) directly more carefully rather than trying to extract it from a coxph object.

But `butcher` absolutely got the job done for my (naive and inefficient) use case.

and yes, lesson learned, I'll try to operate completely in safer tidymodels land next time :)
